mono/utils/mono-flight-recorder.c

```diff
@@ -0,0 +1,153 @@
+/**
+ * \file
+ *     A lightweight log storage medium with limited history
+ *
+ * Author:
+ *   Alexander Kyte (alkyte@microsoft.com)
+ *
+ * (C) 2018 Microsoft, Inc.
+ *
+ */
+
+#include <glib.h>
+#include <mono/utils/mono-flight-recorder.h>
+
+// Need to make lockless because mutex is really slow
+// at each log statement.
+//
+// Better/concurrent design:
+// - Make thread for flight recorder
+// - Send messages to flight recorder using utils/lock-free-queue.h
+// - Flight recorder doesn't share memory, all owned by thread
+
+
+
+#define MONO_FLIGHT_RECORDER_SENTINEL -1
+
+// Mutex has to be held when called
+void
+mono_flight_recorder_iter_init (MonoFlightRecorder *recorder, MonoFlightRecorderIter *iter) 
+{
+	// Make sure we are initialized
+	g_assert (recorder->max_count > 0);
+
+	iter->recorder = recorder;
+	if (recorder->cursor == MONO_FLIGHT_RECORDER_SENTINEL) {
+		iter->lowest_index = MONO_FLIGHT_RECORDER_SENTINEL; 
+		iter->highest_index = MONO_FLIGHT_RECORDER_SENTINEL;
+	} else if (recorder->cursor >= recorder->max_count) {
+		// Ring buffer has wrapped around
+		// So the item *after* the highest index is the lowest index
+		iter->highest_index = (recorder->cursor + 1) % recorder->max_count;
```

Comment:
why `+1`?

Reply:
The cursor points to the last written record.